File: SourceCode/Java_Generic/ChalmersGU/API/SeparateChainingHashMap.java

import java.util.Iterator;

/* *** ODSATag: SeparateChainingHashMap *** */
/* *** ODSATag: Header *** */
class SeparateChainingHashMap<K, V> implements Map<K, V> {
    Map<K, V>[] internalTable;
    int mapSize;

    SeparateChainingHashMap() {
        initialiseTable(MinCapacity);
    }

    @SuppressWarnings("unchecked")
    private void initialiseTable(int capacity) {
        internalTable = (Map<K, V>[]) new LinkedMap[capacity];
        mapSize = 0;
    }

    private Map<K, V> lookupBin(int i) {
        Map<K, V> bin = internalTable[i];
        if (bin == null)
            bin = internalTable[i] = new LinkedMap<>();
        return bin;
    }
/* *** ODSAendTag: Header *** */

/* *** ODSATag: Constants *** */
    static int MinCapacity = 8;
    static double MinLoadFactor = 0.5;
    static double MaxLoadFactor = 2.0;
    static double CapacityMultiplier = 1.5;
/* *** ODSAendTag: Constants *** */

/* *** ODSATag: HashIndex *** */
    private int hash(K key) {
        return (key.hashCode() & 0x7fffffff) % internalTable.length;
    }
/* *** ODSAendTag: HashIndex *** */

/* *** ODSATag: Put *** */
    public V put(K key, V value) {
        int i = hash(key);
        V old = lookupBin(i).put(key, value);
        if (old == null) {
            mapSize++;
            if (loadFactor() > MaxLoadFactor)
                resizeTable((int) (internalTable.length * CapacityMultiplier));
        }
        return old;
    }
/* *** ODSAendTag: Put *** */

/* *** ODSATag: Get *** */
    public V get(K key) {
        int i = hash(key);
        return lookupBin(i).get(key);
    }        
/* *** ODSAendTag: Get *** */

/* *** ODSATag: Remove *** */
    public V remove(K key) {
        int i = hash(key);
        V removed = lookupBin(i).remove(key);
        if (removed != null) {
            mapSize--;
            if (loadFactor() < MinLoadFactor) 
                resizeTable((int) (internalTable.length / CapacityMultiplier));
        }
        return removed;
    }            
/* *** ODSAendTag: Remove *** */

    public boolean containsKey(K key) {
        int i = hash(key);
        return lookupBin(i).containsKey(key);
    }

/* *** ODSATag: Resize *** */
    private void resizeTable(int newCapacity) {
        if (newCapacity < MinCapacity) return;
        Map<K, V>[] oldTable = internalTable;
        initialiseTable(newCapacity);
        for (Map<K, V> bin : oldTable)
            if (bin != null)
                for (K k : bin)
                    put(k, bin.get(k));
    }
/* *** ODSAendTag: Resize *** */

    public boolean isEmpty() {
        return mapSize == 0;
    }

    public int size() {
        return mapSize;
    }

/* *** ODSATag: LoadFactor *** */
    public double loadFactor() {
        return (double) mapSize / internalTable.length;
    }
/* *** ODSAendTag: LoadFactor *** */

    public Iterator<K> iterator() {
        return new HashMapIterator();
    }

    private class HashMapIterator implements Iterator<K> {
        private int bucketIndex;
        private Iterator<K> bucketIterator;
        HashMapIterator() {
            bucketIndex = 0;
            bucketIterator = lookupBin(bucketIndex).iterator();
        }
        public boolean hasNext() {
            while (!bucketIterator.hasNext()) {
                bucketIndex++;
                if (bucketIndex >= internalTable.length) return false;
                bucketIterator = lookupBin(bucketIndex).iterator();
            }
            return true;
        }
        public K next() {
            hasNext();
            return bucketIterator.next();
        }
    }
/* *** ODSAendTag: SeparateChainingHashMap *** */


/***************************************************************************************/
/** What comes below is purely for debugging and testing purposes - it can be removed **/

    public void _printMap() {
        System.out.print(internalTable.length + " [ ");
        for (int i=0; i<internalTable.length; i++) {
            if (i > 0) System.out.print("| ");
            Map<K, V> bin = internalTable[i];
            if (bin == null) System.out.print("? ");
            else if (bin.isEmpty()) System.out.print("- ");
            else for (K k : bin) System.out.print(k + ":" + get(k) + " ");
        }
        System.out.println("] " + size() + " {" + loadFactor() + "}");
    }

    public static void main(String[] args) {
        SeparateChainingHashMap<String, Integer> map = new SeparateChainingHashMap<>();
        map._printMap();
        System.out.println("Putting values");
        for (int i=0; i<40; i++) {
            map.put(String.valueOf((char)((i%25)+65)), i);
            if (i % 5 == 0) map._printMap();
        }
        map._printMap();
        System.out.println("Removing values");
        int i = 7;
        while (!map.isEmpty()) {
            map.remove(String.valueOf((char)((i%25)+65)));
            i += 11;
            if (map.size() % 5 == 0) map._printMap();
        }
    }
/* *** ODSATag: SeparateChainingHashMap *** */
}
/* *** ODSAendTag: SeparateChainingHashMap *** */

File: SourceCode/Python/ChalmersGU/API/SeparateChainingHashMap.py

from API import Map
from LinkedMap import LinkedMap

#/* *** ODSATag: SeparateChainingHashMap *** */
#/* *** ODSATag: Header *** */
class SeparateChainingHashMap(Map):
#/* *** ODSAendTag: Header *** */
#/* *** ODSATag: Constants *** */
    _minCapacity = 8
    _minLoadFactor = 0.5
    _maxLoadFactor = 2.0
    _capacityMultiplier = 1.5
#/* *** ODSAendTag: Constants *** */

#/* *** ODSATag: Constructor *** */
    def __init__(self):
        self._initialiseTable(self._minCapacity)

    def _initialiseTable(self, capacity):
        capacity = int(capacity)
        self._internalTable = [None] * capacity
        for i in range(capacity):
            self._internalTable[i] = LinkedMap()
        self._mapSize = 0
#/* *** ODSAendTag: Constructor *** */

#/* *** ODSATag: HashIndex *** */
    def _hash(self, key):
        return (hash(key) & 0x7fffffff) % len(self._internalTable)
#/* *** ODSAendTag: HashIndex *** */

#/* *** ODSATag: Put *** */
    def put(self, key, value):
        i = self._hash(key)
        old = self._internalTable[i].put(key, value)
        if old is None:
            self._mapSize += 1
            if self.loadFactor() > self._maxLoadFactor:
                self._resizeTable(len(self._internalTable) * self._capacityMultiplier)
        return old
#/* *** ODSAendTag: Put *** */

#/* *** ODSATag: Get *** */
    def get(self, key):
        i = self._hash(key)
        return self._internalTable[i].get(key)
#/* *** ODSAendTag: Get *** */

#/* *** ODSATag: Remove *** */
    def remove(self, key):
        i = self._hash(key)
        removed = self._internalTable[i].remove(key)
        if removed is not None:
            self._mapSize -= 1
            if self.loadFactor() < self._minLoadFactor:
                self._resizeTable(len(self._internalTable) / self._capacityMultiplier)
        return removed
#/* *** ODSAendTag: Remove *** */

    def containsKey(self, key):
        i = self._hash(key)
        return self._internalTable[i].containsKey(key)

#/* *** ODSATag: Resize *** */
    def _resizeTable(self, newCapacity):
        if newCapacity < self._minCapacity: return
        oldTable = self._internalTable
        self._initialiseTable(int(newCapacity))
        for bin in oldTable:
            for k in bin:
                self.put(k, bin.get(k))
#/* *** ODSAendTag: Resize *** */

    def isEmpty(self):
        return self._mapSize == 0

    def size(self):
        return self._mapSize

#/* *** ODSATag: LoadFactor *** */
    def loadFactor(self):
        return self._mapSize / len(self._internalTable)
#/* *** ODSAendTag: LoadFactor *** */

    def __iter__(self):
        for bin in self._internalTable:
            for key in bin:
                yield key
# #/* *** ODSAendTag: SeparateChainingHashMap *** */


#######################################################################################
## What comes below is purely for debugging and testing purposes - it can be removed ##

def _printMap(map):
    print(len(map._internalTable), "[",
          " | ".join(" ".join(f"{k}:{bin.get(k)}" for k in bin) for bin in map._internalTable),
          "]", m.size(), "{", m.loadFactor(), "}")

if __name__ == '__main__':
    m = SeparateChainingHashMap()
    _printMap(m)
    print("Putting values")
    for i in range(40):
        m.put(chr((i%25)+65), i)
        if i % 5 == 0: _printMap(m)
    _printMap(m)
    print("Removing values")
    i = 7
    while not m.isEmpty():
        m.remove(chr((i%25)+65))
        i += 11
        if m.size() % 5 == 0: _printMap(m)
